subsys/bluetooth/host/conn.c

```diff
@@ -1492,33 +1492,46 @@ void bt_conn_unref(struct bt_conn *conn)
 
 	__ASSERT(conn, "Invalid connection reference");
 
-	/* Storing parameters of interest so we don't access the object
-	 * after decrementing its ref-count
+	/* If we're removing the last reference, the connection object will be
+	 * considered freed and possibly reused by the Bluetooth Host stack
+	 * as soon as we decrement the counter.
+	 * To prevent access to such a recycled object, we store the parameters
+	 * of interest before decrementing the reference count,
+	 * then unset the connection pointer.
 	 */
 	conn_type = conn->type;
 	conn_role = conn->role;
 	conn_handle = conn->handle;
-
-	old = atomic_dec(&conn->ref);
-	/* Prevent from accessing connection object */
-	deallocated = (atomic_get(&old) == 1);
 	IF_ENABLED(CONFIG_BT_CONN_TX,
-		   (__ASSERT(!(deallocated && k_work_is_pending(&conn->tx_complete_work)),
-			     "tx_complete_work is pending when conn is deallocated")));
+		   (bool conn_tx_is_pending = k_work_is_pending(&conn->tx_complete_work);));
+	old = atomic_dec(&conn->ref);
 	conn = NULL;
 
 	LOG_DBG("handle %u ref %ld -> %ld", conn_handle, old, (old - 1));
 
 	__ASSERT(old > 0, "Conn reference counter is 0");
 
-	/* Slot has been freed and can be taken. No guarantees are made on requests
-	 * to claim connection object as only the first claim will be served.
+	deallocated = (old == 1);
```

Comment:
```suggestion
	deallocated = (atomic_get(&old) == 1);
```

Given that `old` is a `atomic_val_t`, we should use the appropriate getter for it

Reply:
No, I think you are right here. In fact, `atomic_get` returns an `atomic_val_t`, so my suggestion was fundamentally invalid :D